### PR TITLE
TIP-221: domain layer extraction

### PR DESCRIPTION
**Pull request by jjanvier**
This PR:
* removes the AttributeOptionManager
* unglues attribute type form from its definition without BC break (important for integrators that made a custom attribute type)
* makes all the code (except FileStorage component and bundle) totally agnostic from Flysystem, which is cool :)

Before: 26 coupling rules violated
Now: 18 coupling rules violated, still some work, but we're getting closer :)

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | -
| Changelog updated                 | Y
| Review and 2 GTM                  | 
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Akeneo/Component/FileStorage/FilesystemProvider.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+use League\Flysystem\MountManager;
+
+/**
+ * Resolves a filesystem registered in the MountManager.
+ *
+ * It's a small wrapper to the \League\Flysystem\MountManager to allow the Akeneo and PIM
+ * services to be totally independent/agnostic from Flysystem.
+ *
+ * @author    Julien Janvier <jjanvier@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
```
Comment (skeleton) [inline src/Akeneo/Component/FileStorage/FilesystemProvider.php]: 2016

File: src/Akeneo/Component/FileStorage/FilesystemProvider.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class FilesystemProvider
+{
+    /** @var MountManager */
+    protected $mountManager;
+
+    /**
+     * FilesystemProvider constructor.
```
Comment (skeleton) [inline src/Akeneo/Component/FileStorage/FilesystemProvider.php]: to remove

File: src/Pim/Bundle/CatalogBundle/AttributeType/AttributeTypeInterface.php [hunk 3]
```diff
@@ -4,25 +4,25 @@
...
 use Pim\Component\Catalog\Model\ProductValueInterface;
+use Pim\Component\Catalog\AttributeTypeInterface as BaseAttributeTypeInterface;
 use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\Form\FormInterface;
 
 /**
  * The attribute type interface
  *
+ * TODO: ideally this interface should not be named AttributeTypeInterface but AttributeTypeFormInterface
+ * TODO: and it should not extend extend Pim\Component\Catalog\AttributeTypeInterface
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/AttributeType/AttributeTypeInterface.php]: remove extend

File: src/Pim/Bundle/EnrichBundle/Form/Factory/ProductValueFormFactory.php [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 namespace Pim\Bundle\EnrichBundle\Form\Factory;
 
-use Pim\Bundle\CatalogBundle\AttributeType\AttributeTypeRegistry;
+use Pim\Component\Catalog\AttributeTypeRegistry;
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Form/Factory/ProductValueFormFactory.php]: order

File: src/Pim/Component/Catalog/AttributeTypeInterface.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace Pim\Component\Catalog;
+
+interface AttributeTypeInterface
```
Comment (skeleton) [inline src/Pim/Component/Catalog/AttributeTypeInterface.php]: phpdoc

**Comment by skeleton** [comment]
fix order with php-cs-fixer, fix comments then gtm

**Comment by solivier** [comment]
gtm

File: spec/Pim/Component/Catalog/Validator/ConstraintGuesser/UniqueValueGuesserSpec.php [hunk 6]
```diff
@@ -56,20 +56,20 @@ public function it_does_not_guess_unique_value(AttributeInterface $attribute)
...
-            'entity' => array(AbstractAttributeType::BACKEND_TYPE_ENTITY, false),
-            'integer' => array(AbstractAttributeType::BACKEND_TYPE_INTEGER, true),
-            'media' => array(AbstractAttributeType::BACKEND_TYPE_MEDIA, false),
-            'metric' => array(AbstractAttributeType::BACKEND_TYPE_METRIC, false),
-            'option' => array(AbstractAttributeType::BACKEND_TYPE_OPTION, false),
-            'options' => array(AbstractAttributeType::BACKEND_TYPE_OPTIONS, false),
-            'price' => array(AbstractAttributeType::BACKEND_TYPE_PRICE, false),
-            'text' => array(AbstractAttributeType::BACKEND_TYPE_TEXT, false),
-            'varchar' => array(AbstractAttributeType::BACKEND_TYPE_VARCHAR, true),
+            'boolean' => array(AttributeTypes::BACKEND_TYPE_BOOLEAN, false),
```
Comment (nidup) [inline spec/Pim/Component/Catalog/Validator/ConstraintGuesser/UniqueValueGuesserSpec.php]: add an explaination (maybe in upgrade.md) to document this change

File: src/Akeneo/Component/FileStorage/FilesystemProvider.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ * Resolves a filesystem registered in the MountManager.
+ *
+ * It's a small wrapper to the \League\Flysystem\MountManager to allow the Akeneo and PIM
+ * services to be totally independent/agnostic from Flysystem.
+ *
+ * @author    Julien Janvier <jjanvier@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class FilesystemProvider
```
Comment (nidup) [inline src/Akeneo/Component/FileStorage/FilesystemProvider.php]: yeaaah, much :heart: in general we should wrap any external lib with this kind of provider, for instance, do the same with spout? ping @skeleton @aRn0D @SamirBoulil @willy-ahva

File: src/Pim/Bundle/CatalogBundle/AttributeType/AttributeTypeInterface.php [hunk 8]
```diff
@@ -2,27 +2,27 @@
...
 use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\Form\FormInterface;
 
 /**
  * The attribute type interface
  *
+ * TODO: ideally this interface should not be named AttributeTypeInterface but AttributeTypeFormInterface
+ * TODO: and it should not extend Pim\Component\Catalog\AttributeTypeInterface
+ * TODO: and maybe it should not be present here but in Enrich (to discuss).
+ * TODO: we keep it as is, to avoid major BC breaks for all integrators that built a custom attribute type
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/AttributeType/AttributeTypeInterface.php]: yes, this is the main difficulty, ideally, we should have an attribute type registry (business domain in catalog) and a UIWhateverFormRegistry with other tagged services in Enrich. maybe a way to help this change is allow partners to continue to implem both interfaces and tag twice their attribute type services. in the pim we should separate concerns

File: src/Pim/Component/Catalog/AttributeTypes.php [hunk 9]
```diff
@@ -12,30 +12,34 @@
...
+    const BACKEND_TYPE_DECIMAL = 'decimal';
+    const BACKEND_TYPE_ENTITY = 'entity'; // not used??
+    const BACKEND_TYPE_INTEGER = 'integer'; // not used??
+    const BACKEND_TYPE_MEDIA = 'media';
+    const BACKEND_TYPE_METRIC = 'metric';
+    const BACKEND_TYPE_OPTION = 'option';
+    const BACKEND_TYPE_OPTIONS = 'options';
+    const BACKEND_TYPE_PRICE = 'prices';
+    const BACKEND_TYPE_REF_DATA_OPTION = 'reference_data_option'; // not used??
+    const BACKEND_TYPE_REF_DATA_OPTIONS = 'reference_data_options'; // not used??
```
Comment (nidup) [inline src/Pim/Component/Catalog/AttributeTypes.php]: mark not used as deprecated? or drop them?

**Comment by nidup** [comment]
GTM

**Comment by a2xchip** [comment]
What GTM stands for? @nidup

**Comment by jjanvier** [comment]
@a2xchip it means Good To Merge

File: spec/Akeneo/Component/FileStorage/FilesystemProviderSpec.php [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+namespace spec\Akeneo\Component\FileStorage;
+
+use League\Flysystem\FilesystemInterface;
+use League\Flysystem\MountManager;
+use PhpSpec\ObjectBehavior;
+use Prophecy\Argument;
+
+class FilesystemProviderSpec extends ObjectBehavior
+{
+    function let(MountManager $mountManager)
```
Comment (aRn0D) [inline spec/Akeneo/Component/FileStorage/FilesystemProviderSpec.php]: it miss method 'it_is_initializable'

File: src/Akeneo/Component/FileStorage/FilesystemProvider.php (hunk 7 above)
Comment (aRn0D) [inline src/Akeneo/Component/FileStorage/FilesystemProvider.php]: Yes you should do that for every external lib.

File: spec/Pim/Component/Catalog/Updater/Copier/MediaAttributeCopierSpec.php [hunk 11]
```diff
@@ -23,14 +23,14 @@ function let(
         AttributeValidatorHelper $attrValidatorHelper,
         FileFetcherInterface $fileFetcher,
         FileStorerInterface $fileStorer,
-        MountManager $mountManager
+        FilesystemProvider $filesystemProvider
```
Comment (aRn0D) [inline spec/Pim/Component/Catalog/Updater/Copier/MediaAttributeCopierSpec.php]: File**S**ystemProvider ?

File: spec/Pim/Component/Catalog/Updater/Copier/MediaAttributeCopierSpec.php [hunk 12]
```diff
@@ -78,7 +78,7 @@ function it_copies_when_a_product_value_has_the_values_and_the_media(
         ProductValueInterface $toProductValue,
         FileStorerInterface $fileStorer,
         FileFetcherInterface $fileFetcher,
-        MountManager $mountManager,
+        FilesystemProvider $filesystemProvider,
```
Comment (aRn0D) [inline spec/Pim/Component/Catalog/Updater/Copier/MediaAttributeCopierSpec.php]: You should remove typehint, already done in the constructor.

File: spec/Pim/Component/Catalog/Updater/Copier/MediaAttributeCopierSpec.php [hunk 13]
```diff
@@ -166,7 +166,7 @@ function it_copies_when_a_product_value_has_a_media_but_not_the_target_value(
         FileStorerInterface $fileStorer,
         ProductValueInterface $toProductValue,
         FileFetcherInterface $fileFetcher,
-        MountManager $mountManager,
+        FilesystemProvider $filesystemProvider,
```
Comment (aRn0D) [inline spec/Pim/Component/Catalog/Updater/Copier/MediaAttributeCopierSpec.php]: idem

File: src/Akeneo/Component/FileStorage/FilesystemProvider.php (hunk 7 above)
Comment (aRn0D) [inline src/Akeneo/Component/FileStorage/FilesystemProvider.php]: Should we named that class : File**s**ystemProvider ?

File: spec/Akeneo/Component/FileStorage/FilesystemProviderSpec.php (hunk 10 above)
Comment (willy-ahva) [inline spec/Akeneo/Component/FileStorage/FilesystemProviderSpec.php]: Never understood what is the purpose of `it_is_initializable`

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeOptionController.php [hunk 14]
```diff
@@ -64,10 +68,11 @@ class AttributeOptionController
      * @param FormFactoryInterface         $formFactory
      * @param ViewHandlerInterface         $viewHandler
      * @param AttributeOptionsSorter       $sorter
-     * @param AttributeOptionManager       $optionManager
+     * @param SimpleFactoryInterface       $optionFactory
      * @param SaverInterface               $optionSaver
      * @param RemoverInterface             $optionRemover
      * @param AttributeRepositoryInterface $attributeRepository
+     * @param AttributeOptionRepositoryInterface $attributeOptionRepository
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeOptionController.php]: Could you align doc pls ?

File: spec/Pim/Component/Catalog/Validator/ConstraintGuesser/UniqueValueGuesserSpec.php (hunk 6 above)
Comment (jjanvier) [inline spec/Pim/Component/Catalog/Validator/ConstraintGuesser/UniqueValueGuesserSpec.php]: done

File: src/Akeneo/Component/FileStorage/FilesystemProvider.php (hunk 7 above)
Comment (jjanvier) [inline src/Akeneo/Component/FileStorage/FilesystemProvider.php]: As long as external libs are used in ONE specific component, it no big deal.
The problem here is that Flysystem was used everywhere to retrieve the right filesystem

Comment (jjanvier) [inline src/Akeneo/Component/FileStorage/FilesystemProvider.php]: in all other libs, (SF, FlySystem), it's written Filesystem, so I kept the naming

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeOptionController.php (hunk 14 above)
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeOptionController.php]: add it to cs-fixer rules if you want
personally, I won't ^^

File: src/Pim/Component/Catalog/AttributeTypes.php (hunk 9 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/AttributeTypes.php]: maybe I'm wrong, they are not just used in my catalog I think
I drop the comments

File: spec/Akeneo/Component/FileStorage/FilesystemProviderSpec.php (hunk 10 above)
Comment (jjanvier) [inline spec/Akeneo/Component/FileStorage/FilesystemProviderSpec.php]: when you relocate a file, the spec will crash anyway, so I don't understand the purpose of this method

File: spec/Pim/Component/Catalog/Updater/Copier/MediaAttributeCopierSpec.php (hunk 11 above)
Comment (jjanvier) [inline spec/Pim/Component/Catalog/Updater/Copier/MediaAttributeCopierSpec.php]: nope, explained up

**Pull request merged**